server/queries/MapQueries.cpp

```diff
@@ -157,6 +157,50 @@ CBlockingDialogQuery::CBlockingDialogQuery(CGameHandler * owner, const BlockingD
 	addPlayer(bd.player);
 }
 
+OpenWindowQuery::OpenWindowQuery(CGameHandler * owner, const CGHeroInstance *hero, EOpenWindowMode mode):
+	CDialogQuery(owner),
+	mode(mode)
+{
+	addPlayer(hero->getOwner());
+}
+
+void OpenWindowQuery::onExposure(QueryPtr topQuery)
+{
+	//do nothing - wait for reply
+}
+
+bool OpenWindowQuery::blocksPack(const CPack *pack) const
+{
+	if (mode == EOpenWindowMode::TAVERN_WINDOW)
+	{
+		if(dynamic_ptr_cast<HireHero>(pack) != nullptr)
+			return false;
+	}
+
+	if (mode == EOpenWindowMode::UNIVERSITY_WINDOW)
+	{
+		if(dynamic_ptr_cast<TradeOnMarketplace>(pack) != nullptr)
+			return false;
+	}
+
+	if (mode == EOpenWindowMode::MARKET_WINDOW)
+	{
+		if(dynamic_ptr_cast<ExchangeArtifacts>(pack) != nullptr)
+			return false;
+
+		if(dynamic_ptr_cast<AssembleArtifacts>(pack))
+			return false;
+
+		if(dynamic_ptr_cast<EraseArtifactByClient>(pack))
+			return false;
+
+		if(dynamic_ptr_cast<TradeOnMarketplace>(pack) != nullptr)
+			return false;
```

Comment:
```suggestion
		if(dynamic_ptr_cast<HireHero>(pack))
			return false;
	}

	if (mode == EOpenWindowMode::UNIVERSITY_WINDOW)
	{
		if(dynamic_ptr_cast<TradeOnMarketplace>(pack))
			return false;
	}

	if (mode == EOpenWindowMode::MARKET_WINDOW)
	{
		if(dynamic_ptr_cast<ExchangeArtifacts>(pack))
			return false;

		if(dynamic_ptr_cast<AssembleArtifacts>(pack))
			return false;

		if(dynamic_ptr_cast<EraseArtifactByClient>(pack))
			return false;

		if(dynamic_ptr_cast<TradeOnMarketplace>(pack))
			return false;
```

Reply:
Can we inherit all those packs from one empty class to have only single cast check? Thus, we can simplify adding of new packs in the future

Reply:
Probably not this time.
We also have similar (and even longer) set of checks for exchange dialog.
We also may want to examine content of netpack to allow only packs that affect objects involved in query so simple empty class won't work.